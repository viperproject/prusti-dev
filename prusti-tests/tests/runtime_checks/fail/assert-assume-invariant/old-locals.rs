//@run: 101
use prusti_contracts::*;

fn main() {
    // passes
    test1(&mut 1);
    // passes
    test2(&mut 10);
    // fails:
    test3(&mut 2);
}

#[trusted]
fn test1(x: &mut i32) {
    *x = 2;
    let a = x;
    prusti_assert!(#[insert_runtime_check] old(*a) == 2);
}

#[trusted]
fn test2(x: &mut i32) {
    *x = 1;
    prusti_assert!(#[insert_runtime_check] {let a = x; old(*a) == 1});
}

#[trusted]
fn test3(x: &mut i32) {
    *x = 1;
    // fails: *a is assigned x in old state.
    prusti_assert!(#[insert_runtime_check] old({let a = x; *a == 1}));
}
