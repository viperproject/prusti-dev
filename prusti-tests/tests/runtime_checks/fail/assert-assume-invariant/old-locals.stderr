thread 'main' panicked at $DIR/old-locals.rs:30:44:
Prusti Runtime Checks: Contract prusti_assert!(old({let a = x; *a == 1})) was violated at runtime
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
